library/std/src/sys/wasix/process/process_wasix.rs

```diff
@@ -207,115 +159,173 @@ impl Process {
+            Ok(None)
         }
     }
 }
 
-#[derive(PartialEq, Eq, Clone, Copy)]
-pub struct ExitStatus(c_int);
+#[derive(Clone, Copy)]
+pub struct ExitStatus(JoinStatus);
+
+impl PartialEq<ExitStatus> for ExitStatus {
+    fn eq(&self, other: &Self) -> bool {
+        self.code().eq(&other.code())
+    }
+
+    fn ne(&self, other: &Self) -> bool {
+        self.code().ne(&other.code())
+    }
+}
+
+impl Eq for ExitStatus {}
 
 impl fmt::Debug for ExitStatus {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        f.debug_tuple("wasix_wait_status").field(&self.0).finish()
+        f.debug_tuple("wasix_wait_status")
+            .field(&JoinStatusType::from(self.0.tag).message())
+            .finish()
     }
 }
 
 impl ExitStatus {
-    pub fn new(status: c_int) -> ExitStatus {
+    pub fn new(status: JoinStatus) -> ExitStatus {
         ExitStatus(status)
     }
 
-    fn exited(&self) -> bool {
-        libc::WIFEXITED(self.0)
-    }
-
     pub fn exit_ok(&self) -> Result<(), ExitStatusError> {
-        // This assumes that WIFEXITED(status) && WEXITSTATUS==0 corresponds to status==0. This is
-        // true on all actual versions of Unix, is widely assumed, and is specified in SuS
-        // https://pubs.opengroup.org/onlinepubs/9699919799/functions/wait.html. If it is not
-        // true for a platform pretending to be Unix, the tests (our doctests, and also
-        // procsss_unix/tests.rs) will spot it. `ExitStatusError::code` assumes this too.
-        match NonZero_c_int::try_from(self.0) {
-            /* was nonzero */ Ok(failure) => Err(ExitStatusError(failure)),
-            /* was zero, couldn't convert */ Err(_) => Ok(()),
+        if self.code() == Some(0) {
+            Ok(())
+        } else {
+            Err(ExitStatusError(self.0))
         }
     }
 
     pub fn code(&self) -> Option<i32> {
-        self.exited().then(|| libc::WEXITSTATUS(self.0))
+        unsafe {
+            match JoinStatusType::from(self.0.tag) {
+                wasi::JOIN_STATUS_TYPE_EXIT_NORMAL => Some(self.0.u.exit_normal.raw() as i32),
+                wasi::JOIN_STATUS_TYPE_EXIT_SIGNAL => {
+                    Some(self.0.u.exit_signal.exit_code.raw() as i32)
+                }
+                _ => None,
+            }
+        }
     }
 
     pub fn signal(&self) -> Option<i32> {
-        libc::WIFSIGNALED(self.0).then(|| libc::WTERMSIG(self.0))
+        unsafe {
+            match JoinStatusType::from(self.0.tag) {
+                wasi::JOIN_STATUS_TYPE_STOPPED => Some(self.0.u.stopped.raw() as i32),
+                wasi::JOIN_STATUS_TYPE_EXIT_SIGNAL => {
+                    Some(self.0.u.exit_signal.signal.raw() as i32)
+                }
+                _ => None,
+            }
+        }
     }
 
     pub fn core_dumped(&self) -> bool {
-        libc::WIFSIGNALED(self.0) && libc::WCOREDUMP(self.0)
+        self.code().map(|c| libc::WIFSIGNALED(c) && libc::WCOREDUMP(c)).unwrap_or(false)
```

Comment:
I'm not sure if it makes sens to call `libc::WIFSIGNALED`/`libc::WCOREDUMP` on a value returned from `proc_spawn`

Reply:
Looks sensible to me.